storage/src/sparse_merkle/internal_node.cpp

```diff
@@ -0,0 +1,171 @@
+// Concord
+//
+// Copyright (c) 2019 VMware, Inc. All Rights Reserved.
+//
+// This product is licensed to you under the Apache 2.0 license (the "License").
+// You may not use this product except in compliance with the Apache 2.0 License.
+//
+// This product may include a number of subcomponents with separate copyright
+// notices and license terms. Your use of these subcomponents is subject to the
+// terms and conditions of the sub-component's license, as noted in the
+// LICENSE file.
+
+#include "sparse_merkle/internal_node.h"
+
+namespace concord {
+namespace storage {
+namespace sparse_merkle {
+
+void BatchedInternalNode::update_hashes(size_t index, Version version) {
+  Assert(index > 0);
+  auto hasher = Hasher();
+
+  while (true) {
+    Hash parent_hash;
+    if (is_left_child(index)) {
+      parent_hash = hasher.parent(get_hash(index), get_hash(index + 1));
+    } else {
+      parent_hash = hasher.parent(get_hash(index - 1), get_hash(index));
+    }
+    auto parent_idx = parent_index(index).value();
+    auto& parent = children_.at(parent_idx);
+    if (!parent) {
+      // We are adding the first child below a non-existent parent. Create the
+      // parent first.
+      children_[parent_idx] = InternalChild{Hash(), version};
+      parent = children_.at(parent_idx);
+    }
+    // A parent is *always* an InternalChild
+    auto& internal_child = std::get<InternalChild>(parent.value());
+    internal_child.hash = parent_hash;
+    internal_child.version = version;
+    if (parent_idx == 0) {
+      // We have reached the root of this BatchedInternalNode
+      return;
+    }
+    index = parent_idx;
+  }
+}
+
+BatchedInternalNode::InsertResult BatchedInternalNode::insert(LeafChild& child, size_t depth) {
```

Comment:
Can we make the `child` parameter a const ref -`const LeafChild&`? That also propagates to `insert_into_empty_child()`, `overwrite()` and `split_node()`.

Reply:
I missed @yuliasherman 's comment above on `insert_into_empty_child()`. This comment adds to it.

Reply:
Good catch. Fixed.